### PR TITLE
CFE-2850: Added function string_replace

### PR DESCRIPTION
**Pull request by karlhto**
Spent a lot more time doing this than necessary because I was trying to create a Buffer function for doing simple replacement, but didn't check string_lib until later.

Wanted to open the PR quickly in case I should use buffers instead, so I haven't pushed tests yet. Will do soon.

### CODE REVIEW

**Comment by karlhto** [comment]
Maybe options for replacing only once or a given number of times should be added?

File: libpromises/evalfunction.c [hunk 1]
```diff
@@ -7062,6 +7062,27 @@ static FnCallResult FnCallStringSplit(ARG_UNUSED EvalContext *ctx,
...
+                                        ARG_UNUSED const FnCall *fp,
+                                        const Rlist *finalargs)
+{
+    char *string = RlistScalarValue(finalargs);
+    char *match = RlistScalarValue(finalargs->next);
+    char *substitute = RlistScalarValue(finalargs->next->next);
+
+    char *ret = SearchAndReplace(string, match, substitute);
+
+    if (!ret)
```
Comment (vpodzime) [inline libpromises/evalfunction.c]: ``ret == NULL``

File: libpromises/evalfunction.c [hunk 2]
```diff
@@ -8779,6 +8800,14 @@ static const FnCallArg STRFTIME_ARGS[] =
     {NULL, CF_DATA_TYPE_NONE, NULL}
 };
 
+static const FnCallArg STRING_REPLACE_ARGS[] =
+{
+    {CF_ANYSTRING, CF_DATA_TYPE_STRING, "Source string"},
+    {CF_ANYSTRING, CF_DATA_TYPE_STRING, "String to search for"},
```
Comment (vpodzime) [inline libpromises/evalfunction.c]: *String to replace*

File: libpromises/evalfunction.c [hunk 3]
```diff
@@ -9276,6 +9305,8 @@ const FnCallType CF_FNCALL_TYPES[] =
                   FNCALL_OPTION_COLLECTING|FNCALL_OPTION_VARARG, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
     FnCallTypeNew("string_split", CF_DATA_TYPE_STRING_LIST, SPLITSTRING_ARGS, &FnCallStringSplit, "Convert a string in arg1 into a list of at most arg3 strings by splitting on a regular expression in arg2",
                   FNCALL_OPTION_NONE, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
+    FnCallTypeNew("string_replace", CF_DATA_TYPE_STRING, STRING_REPLACE_ARGS, &FnCallStringReplace, "Replace occurrences of arg1 in arg2 with arg3.",
```
Comment (vpodzime) [inline libpromises/evalfunction.c]: The descriptions of the args above do not match the explanation here.

Comment (vpodzime) [reply libpromises/evalfunction.c]: `arg1` and `arg2` are swapped in this description.

File: tests/acceptance/01_vars/02_functions/string_replace.cf [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+bundle agent test
+{
+  vars:
+      "test" string => "abcdefghij";
+
+      # replace the empty string with empty string, globally
+      "void" string => string_replace($(test), "", "");
+      # replace the empty string with 'x', globally (matches on character boundaries)
+      "void1" string => string_replace($(test), "", "x");
```
Comment (vpodzime) [inline tests/acceptance/01_vars/02_functions/string_replace.cf]: I don't really understand what this is supposed to do.

File: tests/acceptance/01_vars/02_functions/string_replace.cf [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+#######################################################
+
+bundle agent test
+{
+  vars:
+      "test" string => "abcdefghij";
+
+      # replace the empty string with empty string, globally
+      "void" string => string_replace($(test), "", "");
+      # replace the empty string with 'x', globally (matches on character boundaries)
```
Comment (vpodzime) [inline tests/acceptance/01_vars/02_functions/string_replace.cf]: What does *globally* mean here? I guess it's just a copy-paste from some other test?

File: tests/acceptance/01_vars/02_functions/string_replace.cf [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      "void1" string => string_replace($(test), "", "x");
+
+      # no matches, original intact
+      "nomatch" string => string_replace($(test), "nonesuch", "TEST");
+
+      # empty cases
+      "empty1" string => string_replace("", "abc", "ABC");
+      "empty2" string => string_replace("", "abc", "");
+}
```
Comment (vpodzime) [inline tests/acceptance/01_vars/02_functions/string_replace.cf]: What about some tests that actually test the functionality not just the wrong cases?

File: tests/acceptance/01_vars/02_functions/string_replace.cf [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+#######################################################
+
+bundle agent check
+{
+  methods:
+      "check"  usebundle => dcs_check_state(test,
+                                           "$(this.promise_filename).expected.json",
+                                           $(this.promise_filename));
```
Comment (vpodzime) [inline tests/acceptance/01_vars/02_functions/string_replace.cf]: Weird alignment here.

File: tests/acceptance/01_vars/02_functions/string_replace.cf (hunk 5 above)
Comment (karlhto) [reply tests/acceptance/01_vars/02_functions/string_replace.cf]: Yep, it's from regex_replace. Globally means that it adds the replacement character between every character. The reason I included it in the tests is if we want that functionality.

Comment (vpodzime) [reply tests/acceptance/01_vars/02_functions/string_replace.cf]: But there's no such argument for the function that would enable/disable this, right? So I think it's quite confusing in the comment.

File: libpromises/evalfunction.c [hunk 8]
```diff
@@ -7062,6 +7062,27 @@ static FnCallResult FnCallStringSplit(ARG_UNUSED EvalContext *ctx,
 
 /*********************************************************************/
 
+static FnCallResult FnCallStringReplace(ARG_UNUSED EvalContext *ctx,
+                                        ARG_UNUSED const Policy *policy,
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: `const` applies to the left, unless it is the first word, then it applies to the right. Since you're using the ARG_UNUSED macro here, I think it might be more clear to move it past `Policy`. Not a big deal, this should work.

File: libpromises/evalfunction.c [hunk 9]
```diff
@@ -7062,6 +7062,27 @@ static FnCallResult FnCallStringSplit(ARG_UNUSED EvalContext *ctx,
...
+                                        ARG_UNUSED const FnCall *fp,
+                                        const Rlist *finalargs)
+{
+    char *string = RlistScalarValue(finalargs);
+    char *match = RlistScalarValue(finalargs->next);
+    char *substitute = RlistScalarValue(finalargs->next->next);
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: Shouldn't you check for NULL before dereferencing pointers? At least an assert or two. Have you tried using `string_replace` with too few args?

File: libpromises/evalfunction.c [hunk 10]
```diff
@@ -7062,6 +7062,27 @@ static FnCallResult FnCallStringSplit(ARG_UNUSED EvalContext *ctx,
...
+
+    char *ret = SearchAndReplace(string, match, substitute);
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: What does `SearchAndReplace` do if `match` is empty string? Personally, I think it should match nowhere and replace nothing.

File: libpromises/evalfunction.c [hunk 11]
```diff
@@ -9276,6 +9305,8 @@ const FnCallType CF_FNCALL_TYPES[] =
                   FNCALL_OPTION_COLLECTING|FNCALL_OPTION_VARARG, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
     FnCallTypeNew("string_split", CF_DATA_TYPE_STRING_LIST, SPLITSTRING_ARGS, &FnCallStringSplit, "Convert a string in arg1 into a list of at most arg3 strings by splitting on a regular expression in arg2",
                   FNCALL_OPTION_NONE, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
+    FnCallTypeNew("string_replace", CF_DATA_TYPE_STRING, STRING_REPLACE_ARGS, &FnCallStringReplace, "Replace occurrences of arg2 in arg1 with arg3.",
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: I would write this so the args are in correct order;

> Search through arg1, replacing occurences of arg2 with arg3.

File: tests/acceptance/01_vars/02_functions/string_replace.cf [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+#######################################################
+
+bundle agent test
+{
+  vars:
+      "test" string => "abcdefghij\t\n";
+      "test2" string => "(){}[].*?";
+
```
Comment (olehermanse) [inline tests/acceptance/01_vars/02_functions/string_replace.cf]: I would add some **really** basic test cases here first. It's great that you're testing special characters and edge cases, but if someone new is reading this test it is easier if you start with the "normal usage".

```
	"match_once" string => string_replace("abcdef", "abc", "ABC");
    "match_twice" string => string_replace("abcdabcd", "abcd", "hello");
    "match_none" string => string_replace("abdc", "abcd", "wrong");
```

And then, of course, compare results.

File: tests/acceptance/01_vars/02_functions/string_replace.cf [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      # normal tests
+      "replace" string => string_replace($(test), "\t", "\r");
+      "replace2" string => string_replace($(test), "ghij\t", "tEsT");
+      "replace3" string => string_replace($(test), "abcdefghij\t\n", "no");
+
```
Comment (olehermanse) [inline tests/acceptance/01_vars/02_functions/string_replace.cf]: What happens if there is an overlap?

```
    "being_stupid" string => string_replace("aaaa", "aaa", "hello");
```

I hope it skips what has already been matched and results in `"helloa"`?

**Comment by olehermanse** [comment]
@karlhto After this PR is done, you should add an example in `core/examples`, and then use that example to write an entry in documentation. The examples are also automatically run as tests.

**Comment by olehermanse** [comment]
@karlhto The acceptance test is currently failing in travis, and needs to be fixed.

File: libpromises/evalfunction.c (hunk 10 above)
Comment (karlhto) [reply libpromises/evalfunction.c]: That's what it does, in contrast to regex_replace, which can replace the empty string.

**Comment by karlhto** [comment]
@cf-bottom Jenkins, please!

**Comment by karlhto** [comment]
@olehermanse @vpodzime Should be fine now :)

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/906/)

https://ci.cfengine.com/job/pr-pipeline/906/

**Comment by vpodzime** [comment]
A build including the exotics:
[Build Status](https://ci.cfengine.com/job/pr-pipeline/909/)

File: libpromises/evalfunction.c [hunk 14]
```diff
@@ -7062,6 +7062,30 @@ static FnCallResult FnCallStringSplit(ARG_UNUSED EvalContext *ctx,
...
+                                        ARG_UNUSED const Policy *policy,
+                                        ARG_UNUSED const FnCall *fp,
+                                        const Rlist *finalargs)
+{
+    char *string = RlistScalarValue(finalargs);
+    assert(string != NULL);
+    char *match = RlistScalarValue(finalargs->next);
+    assert(match != NULL);
+    char *substitute = RlistScalarValue(finalargs->next->next);
+    assert(substitute != NULL);
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: What happens if you give an incorrect number of args?

File: libpromises/evalfunction.c (hunk 9 above)
Comment (olehermanse) [reply libpromises/evalfunction.c]: There is already code which should check that arguments are correct, and ensure that `finalargs->next->next` is valid for this function. But being extra cautios and adding ifs is also fine.

File: libpromises/evalfunction.c (hunk 14 above)
Comment (olehermanse) [reply libpromises/evalfunction.c]: Please test this (it shouldn't segfault or abort).

**Comment by olehermanse** [comment]
@karlhto Very close to done. Please respond to my comments and test what I suggested. Converting the asserts to ifs might be a good idea.

**Comment by karlhto** [comment]
¡@cf-bottom jenkins!

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/937/)

https://ci.cfengine.com/job/pr-pipeline/937/

**Comment by karlhto** [comment]
@cf-bottom jenkins?

**Comment by karlhto** [comment]
@cf-bottom jenkins!!

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/966/)

https://ci.cfengine.com/job/pr-pipeline/966/

**Comment by karlhto** [comment]
@olehermanse @vpodzime One final review?

File: libpromises/evalfunction.c [hunk 15]
```diff
@@ -7062,6 +7062,39 @@ static FnCallResult FnCallStringSplit(ARG_UNUSED EvalContext *ctx,
...
 /*********************************************************************/
 
+static FnCallResult FnCallStringReplace(ARG_UNUSED EvalContext *ctx,
+                                        ARG_UNUSED const Policy *policy,
+                                        ARG_UNUSED const FnCall *fp,
+                                        const Rlist *finalargs)
+{
+    if (finalargs->next == NULL || finalargs->next->next == NULL)
+    {
+        Log(LOG_LEVEL_VERBOSE,
```
Comment (vpodzime) [inline libpromises/evalfunction.c]: I think this should be a warning.

File: libpromises/evalfunction.c [hunk 16]
```diff
@@ -7062,6 +7062,39 @@ static FnCallResult FnCallStringSplit(ARG_UNUSED EvalContext *ctx,
...
+
+    char *string = RlistScalarValue(finalargs);
+    char *match = RlistScalarValue(finalargs->next);
+    char *substitute = RlistScalarValue(finalargs->next->next);
+
+    char *ret = SearchAndReplace(string, match, substitute);
+
+    if (ret == NULL)
+    {
+        Log(LOG_LEVEL_VERBOSE,
```
Comment (vpodzime) [inline libpromises/evalfunction.c]: Same here.

**Comment by karlhto** [comment]
@cf-bottom @jenkins

**Comment by karlhto** [comment]
@cf-bottom jenkins pls

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/970/)

https://ci.cfengine.com/job/pr-pipeline/970/

**Comment by karlhto** [comment]
@vpodzime I think it's ready to merge now :)

**Comment by vpodzime** [comment]
> @vpodzime I think it's ready to merge now :)

Absolutely!

**Comment by vpodzime** [comment]
@karlhto could you please cherry-pick this to the *3.12.x* and *3.10.x* branches? Also, could you please submit PRs for documentation?

**Comment by olehermanse** [comment]
Good job @karlhto.
